lib/generators/logidze/install/check_pending.rb

```diff
@@ -0,0 +1,45 @@
+# frozen_string_literal: true
+
+require_relative "function_definitions"
+
+module Logidze
+  class PendingMigrationError < StandardError
+  end
+
+  module Generators
+    class CheckPending
+      def initialize(app)
+        @app = app
+      end
+
+      delegate :connection, to: ActiveRecord::Base
+
+      def call(env)
+        notify_or_raise! if Logidze.check_pending_upgrade && needs_migration?
+
+        @app.call(env)
+      end
+
+      private
+
+      def notify_or_raise!
+        msg = "Logidze needs upgrade. Run `bundle exec rails generate logidze:install --update`"
+        raise Logidze::PendingMigrationError, msg if Logidze.raise_on_pending_upgrade
+
+        warn msg
+      end
+
+      def needs_migration?
+        (library_function_versions - pg_function_versions).any?
+      end
+
+      def pg_function_versions
+        Logidze::Generators::FunctionDefinitions.from_db.map { |func| [func.name, func.version] }
```

Comment:
Calling this function for every request could be costly; we need some kind of memoization here as well.

Active Record, for example, watches file changes: https://github.com/rails/rails/blob/098fd7f9b3d5c6f540911bc0c17207d6b48d5bb3/activerecord/lib/active_record/migration.rb#L578

Reply:
Sure! But when should we invalidate cache with the "warn" option?
With the "error" option it's easy - only cache results on success. But warnings can be called infinitely without any notification for the end user.

With this in mind, is there really any need for the "warn" option or should we drop it?

Reply:
> warnings can be called infinitely without any notification for the end user.

If we implement files watching, we should only perform the check when something changed, so, it won't be too often.

Reply:
We can watch the migrations folder for updates and then clear the cache, but I have one edge case in mind. Let's say I use the "error" option and do not check for migrations. I can run 'logidze update' without migrations, and then on the next request, an error will be cached until the next migration file is created, or server reload.

Without the 'warn' option, we can cache only success, just like it is implemented in AR. We can also do not watch for file updates because gem update implies server reloading.

Or may I get it wrong?

Reply:
> I can run 'logidze update' without migrations

What do you mean?

With file watcher, we shouldn't cache anything. Imaging the pseudo-code:

```ruby
def call
   @watcher ||= build_watcher do
      @needs_check = true
      Logidze.validate_version! # this line either raises an exception, thus making the next line no-op or prints a warning
      @needs_check = false
   end
   
   if @needs_check
     watcher.execute # calls the block above
   else
     watcher.execute_if_updated # calls the block above only if nothing has changed
   end
end
```

Initially, `@needs_check` is set to true, so, the first request always calls `validate_version!`.

If versions match or the `warn` mode is used, `@needs_check` is set to false, thus,  subsequent requests always call `#execute_if_updated`—which, as it states, only calls the block if the watched files were modified.

If versions do not match, we perform the check and raise an exception for every request until this is fixed.

Reply:
I see, but if the warning is printed only once for a while, why do we need file checks at all?